STACpopulator/extensions/base.py

```diff
@@ -0,0 +1,331 @@
+        item_ext = ExtSubCls.ext(item, add_if_missing=add_if_missing)
+        item_ext.apply(self.model_dump(mode="json", by_alias=True))
+        return item
+
+    def to_stac_schema(self) -> dict:
+        """Return the STAC schema for the extension."""
+        return {'type': 'object',
+                'required': ['type', 'properties'],
+                'properties': {'type': {'const': 'Feature'},
+                               'properties': {'$ref': str(self._schema_uri)}
+                               }
+                }
+    def write_stac_schema(self) -> str:
+        path = f"/tmp/{self._prefix}-schema.json"
+        with open(path, "w") as fh:
+            json.dump(self.to_stac_schema(), fh)
+        return path
+
+
+class BaseSTAC(BaseModel):
+    """Base class for STAC item data models.
+
+    Attributes
+    ----------
+    geometry : AnyGeometry
+        The geometry of the dataset.
+    bbox : list[float]
+        The bounding box of the dataset.
+    start_datetime : datetime
+        The start datetime of the dataset.
+    end_datetime : datetime
+        The end datetime of the dataset.
+    extensions : list[str]
+        Name of the class attributes that point to STAC extension helper classes. Those extension classes should have an `apply` method.
+    """
+    # STAC item properties
+    geometry: AnyGeometry | None
+    bbox: list[float]
+    start_datetime: datetime
+    end_datetime: datetime
+
+    model_config = ConfigDict(populate_by_name=True, extra="ignore", arbitrary_types_allowed=True)
+
+    # Helpers are automatically detected by being Helper subclasses
+    _helpers: list[str] = PrivateAttr([])
+
+    @property
+    def uid(self) -> str:
+        """Return a unique ID. When subclassing, use a combination of properties uniquely identifying a dataset."""
+        # TODO: Should this be an abstract method?
+        import uuid
+        return str(uuid.uuid4())
+
+    @model_validator(mode="after")
+    def find_helpers(self):
+        """Populate the list of extensions."""
+        for key, field in self.model_fields.items():
+            if isinstance(field.annotation, type) and issubclass(field.annotation, Helper):
+                self._helpers.append(key)
+
+    def stac_item(self) -> "pystac.Item":
+        """Create a STAC item and add extensions."""
+        item = pystac.Item(
+            id=self.uid,
+            geometry=self.geometry.model_dump(),
+            bbox=self.bbox,
+            start_datetime=self.start_datetime,
+            end_datetime=self.end_datetime,
+            datetime=None,
+            properties={},
+        )
+
+        # Add extensions
+        for ext in self._helpers:
+            getattr(self, ext).apply(item)
+
+        try:
+            item.validate()
+        except STACValidationError as e:
+            raise Exception("Failed to validate STAC item") from e
+
+        return json.loads(json.dumps(item.to_dict()))
```

Comment:
Why `dumps` and then `loads`? Do we not trust that `to_dict` will be correct?

Reply:
But if there's problems with serialization won't `dumps` fail?

Or is pystac using a different (worse?) serialization library than json?

I'm only balking at this because this seems like a hacky solution to something that should be handled properly by the pystac library.... 

Do you remember what values couldn't be serialized? If so, can we add an issue on pystac and add a comment here about it?

Reply:
I recall this as well. 

I think it was due to some string-like nested objects (`href` or `media_type` in links/assets, not sure, it's been a while...) not getting converted to valid JSON with `to_dict` directly, and those were not nicely handled by the code receiving them (like `requests` not liking the items to POST or similar). 

It's definitely been a while and many `pystac` updates since, so might be worth a revisit.